akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -306,6 +312,12 @@ private[persistence] trait Eventsourced extends Snapshotter with Stash with Stas
    * @param handler handler for each persisted `event`
    */
   final def persistAsync[A](event: A)(handler: A ⇒ Unit): Unit = {
+    if (insidePersistCallback) {
+      val ex = new UnsupportedOperationException("calling persistAsync from within the context of a persist/persistAsync callback is not supported!")
+      try onPersistFailure(ex, event, sequenceNr) finally context.stop(self)
```

Comment:
yes, it should behave exactly as when exc from ordinary receive, and that is why we have those finally and catch and things to make sure the state is correct (especially important for resume)